hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java

```diff
@@ -631,4 +647,40 @@ private static boolean hiveEngineEnabled(TableMetadata metadata, Configuration c
 
     return conf.getBoolean(ConfigProperties.ENGINE_HIVE_ENABLED, TableProperties.ENGINE_HIVE_ENABLED_DEFAULT);
   }
+
+  private static class HiveLockHeartbeat implements Runnable {
+    private final ClientPool<IMetaStoreClient, TException> hmsClients;
+    private final long lockId;
+    private final long intervalMs;
+    private ScheduledFuture<?> future;
+
+    HiveLockHeartbeat(ClientPool<IMetaStoreClient, TException> hmsClients, long lockId, long intervalMs) {
+      this.hmsClients = hmsClients;
+      this.lockId = lockId;
+      this.intervalMs = intervalMs;
+      this.future = null;
+    }
+
+    @Override
+    public void run() {
+      try {
+        hmsClients.run(client -> {
+          client.heartbeat(0, lockId);
+          return null;
+        });
+      } catch (TException | InterruptedException e) {
+        LOG.error("Fail to heartbeat for lock: {}", lockId, e);
```

Comment:
Thanks, I also think we should fail the commit on heart beat failure. However, to do so we would have to add retry on heart beats as well, so that transient failures don't fail the entire job. I will add that.

For the unlock failure handling, I think that needs a bit more discussion and if it is OK with you let's do that discussion in a follow up PR.

Reply:
RetryingMetastoreClient does the retry for us. No need to do the retries by hand

Reply:
Oh yea, forgot about the niceties of `RetryingMetastoreClient`. Thanks!

Reply:
@SinghAsDev: We can do some elaborate stuff to throw `CommitFailedException` if we are before the `alter` operation, and we can throw `CommitStateUnknownException` if we called the `alter` command already, but I hope this would be a rare occurrence, so it might not worth the extra complexity. The question is how we will be able to communicate back to the original thread that the heartbeat is failed. Based on this the decision above could be an easy one, or a more complex one, and then we can decide which one to chose.

Reply:
@pvary I agree that this added complexity does not really buy us much. All the lengthy operations I have seen that needed heartbeats are within the alter table operation, so very unlikely we will see heartbeat failure before alter operation. After alter table is done, any issue with lock, we will anyway see in the unlock operation. As such, I still think we likely don't need the added complexity. If we agree on this reasoning, I can add this info in the code as well.

Reply:
Ok. So we would likely to need a test where the failing heartbeat will fail the commit with the `CommitStateUnknownException`. So It is fine if we have `CommitStateUnknownException` for all of these cases

Reply:
Added a test. Can you help take a look @pvary @szehon-ho , thanks!